### PR TITLE
Add TorchScript lazy tensor backend

### PR DESCRIPTION
**Pull request by asuhan**
Minimal backend using lay tensors, with tests imported from PT/XLA.

Test Plan:

test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.*
python example.py

### CODE REVIEW

**Comment by asuhan** [comment]
The interesting code is under `ts_backend`.

Tests have been imported from pytorch/xla with only minimal adjustments, mostly around disabling a few which we'll fix later. Other changes have been required for build scripts and patching the existing code generation infrastructure in PyTorch.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      graph_(std::make_shared<torch::jit::Graph>()) {}
+
+TSLoweringContext::TSLoweringContext(
+    const std::string& name, Device device,
+    absl::Span<const ir::Node* const> post_order,
+    ir::Util::EmissionMap emit_status)
+    : ir::LoweringContext(name, device, post_order, emit_status),
+      graph_(std::make_shared<torch::jit::Graph>()) {
+  auto lowering = NodeLowering::Create(this);
+  for (auto node : post_order) {
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.cpp:20]: why not use LowerNode(node) as below?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  GenericComputationTS(std::shared_ptr<torch::jit::Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  lazy_tensors::StatusOr<lazy_tensors::ProgramShape> GetProgramShape()
+      const override {
+    std::vector<std::string> parameter_names;
+    for (torch::jit::Value* input : graph_->inputs()) {
+      parameter_names.push_back(input->debugName());
+    }
+    // TODO(asuhan): should we use valid shapes?
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.h:25]: what does this mean?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+      return false;
+    }
+    LTC_CHECK_EQ(node->num_outputs(), ops.size());
+    for (size_t i = 0; i < ops.size(); ++i) {
+      loctx()->AssignOutputOp(ir::Output(node, i), ops[i]);
+    }
+    return true;
+  }
+
+  lazy_tensors::Shape Infer(const ir::Node* node) override {
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:46]: oh man.. 
so, first how confident are you about test coverage of what you have here?
second, what does this look like in the world we can use @bdhirsh's work with meta-kernels?

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      std::vector<lazy_tensors::int64> padded_dimensions(
+          argument_dimensions.begin(), argument_dimensions.end());
+      size_t i = 0;
+      for (auto rit = pad.rbegin(); rit != pad.rend(); rit += 2, ++i) {
+        padded_dimensions[i] += (*rit + *(rit + 1));
+      }
+      return lazy_tensors::Shape(argument_shape.element_type(),
+                                 padded_dimensions);
+    }
+    LTC_LOG(FATAL) << *node << "Not implemented yet.";
```
Comment (wconstab) [inline lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:78]: i'm not clear on which cases Infer even gets called.  where do I look to understand that?

**Review by wconstab — APPROVED** [review]
I looked over everything in detail, except only skimming the lowerings, xla_type, and tests.

Anything in particular you want me to check more carefully? 

I think it looks good.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  lazy_tensors::ComputationClient* GetComputationClientIfInitialized()
+      const override {
+    return lazy_tensors::compiler::TSClientGetIfInitialized();
+  }
+
+  std::vector<std::string> GetCompilationDevices(
+      const std::string& device,
+      lazy_tensors::Span<const std::string> devices) const override {
+    return std::vector<std::string>(devices.begin(), devices.end());
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp:44]: This is related to replication support for TPUs, it's identity for GPUs and CPUs. IIRC this avoids compiling the same program for all the cores in a slice in data parallelism scenarios, it wouldn't surprise me if future accelerators follow this model too.

File: lazy_tensor_core/example.py [hunk 6]
```diff
@@ -2,6 +2,9 @@
 import lazy_tensor_core
 import lazy_tensor_core.debug.metrics as metrics
 
+lazy_tensor_core._LAZYC._ltc_init_ts_backend()
+lazy_tensor_core._LAZYC._initialize_aten_bindings()
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/example.py:6]: We don't want to register the TorchScript backend by default, I don't think re-registering a different backend is guaranteed to be correct in the PyTorch core.

File: lazy_tensor_core/test/cpp/torch_ltc_ts_test.cpp [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+#include <ATen/ATen.h>
+
+#include "lazy_tensor_core/csrc/tensor.h"
+
+namespace torch_lazy_tensors {
+namespace cpp_test {
+
+void LtcTsTest::SetUp() {
+  at::manual_seed(42);
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/test/cpp/torch_ltc_ts_test.cpp:11]: Yes.

File: lazy_tensor_core/torch_patches/X10-codegen.diff [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+diff --git a/tools/codegen/dest/gen_external_aten_fallbacks.py b/tools/codegen/dest/gen_external_aten_fallbacks.py
+index ee4a2c30d6..b516ebac8f 100644
+--- a/tools/codegen/dest/gen_external_aten_fallbacks.py
++++ b/tools/codegen/dest/gen_external_aten_fallbacks.py
+@@ -30,40 +30,7 @@ _FN_DENYLIST_REGEX = [
+ # Instead, the codegen will figure out which ops to generate _out wrappers for
+ # entirely from the yaml. Maintaining the same behavior as current XLA codegen for now.
+ _FN_OUT = [
+-    'abs',
+     'add',
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/torch_patches/X10-codegen.diff:48]: We leave `add` alone here, we don't remove it. We lower it.

[the remark this answers is not in the document]
Comment (bdhirsh) [reply lazy_tensor_core/torch_patches/X10-codegen.diff:48]: I'm sure I'm missing a bunch of a context, but just FYI, you're gonna get some merge conflicts once https://github.com/pytorch/pytorch/pull/58065 lands :)

In that PR I'm deleting this file, and replacing all of the code-generated CPU fallback kernels with a single boxed CPU fallback.

We can also talk more about the (potential) integration with the codegen will look like in sync tomorrow.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_computation_client.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    bool HasValue() const override { return data_.defined(); }
+
+    at::Tensor data_;
+  };
+
+  DataPtr CreateDataPlaceholder(std::string device, Shape shape) override;
+
+  std::vector<DataPtr> TransferToServer(
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_computation_client.h:34]: They're necessary, it's just that they occur in patterns more complicated than ones exposed by unit tests I've already added. One of the more obvious uses is here: https://github.com/pytorch/pytorch/blob/0e8776b4d45a243df6e8499d070e2df89dcad1f9/lazy_tensor_core/lazy_tensor_core/core/lazy_model.py#L840, you can track it all the way to this method.

I didn't want to add an implementation for which we don't have a test yet, as we add them back all these issues will be exposed.

File: lazy_tensor_core/ts_native_functions.yaml [hunk 10]
```diff
@@ -0,0 +1,34 @@
+backend: XLA
```
[the remark this answers is not in the document]
Comment (bdhirsh) [reply lazy_tensor_core/ts_native_functions.yaml:1]: This isn't gonna work out of the box- in addition to generating some boilerplate, the codegen also tries to register a bunch of stuff to the dispatcher, and it expects the passed-in backend to be a valid DispatchKey that it can register to.

We can talk more about what the codegen bits of the integration should look like at the sync. I probably need to look more at the LTC branch to see what it's doing, but it sounds like the codegen will be useful for generating most of the boilerplate, minus any dispatcher registrations.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.h (hunk 2 above)
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.h:25]: The program shape could be used by the compiler back-end for various purposes, such as assigning a certain physical layout - that's what happens with the TPU back-end, for example. I've checked and I'm pretty sure we're not using anywhere else, so for this particular back-end it shouldn't matter. I think I've answered my own TODO and I should add this observation instead - we don't need the actual result shape here.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 3 above)
Comment (bdhirsh) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:46]: side note- I'm hoping that as we get better meta coverage, we'll be able to compute shapes on the spot for **all** xla ops, and finally de-virtualize `Tensor::sizes()`. That doesn't seem worth doing right now in `pytorch/xla`, but seems a lot more reasonable once meta coverage improves and to do directly on LazyTensor (versus trying to do it to the `XLATensor` subclass that's currently in pytorch/xla right now, which would result in a bunch of redundant work).

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.cpp (hunk 1 above)
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_lowering_context.cpp:20]: It was a bad helper, I've removed it instead. It instantiated a new node lowering every time, which isn't a semantic problem since the lowering doesn't have state but still a janky way to do it.

File: lazy_tensor_core/setup.py [hunk 11]
```diff
@@ -100,6 +100,15 @@ def create_version_files(base_dir, version, ltc_git_sha, torch_git_sha):
     f.write('}  // namespace torch_lazy_tensors\n')
 
 
+def generate_xla_aten_code(base_dir):
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/setup.py]: Good catch, fixed.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp (hunk 3 above)
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp:46]: @bdhirsh Faster dispatch is always a good thing! I'm not too worried about it anyway, we cache inference results, so it's not the end of the world if it's not running at lightspeed.

File: lazy_tensor_core/ts_native_functions.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/ts_native_functions.yaml:1]: @bdhirsh I agree & was planning to discuss this topic during our meeting. We should de-hardcode names / namespaces and not force backends to lower all operations which have `_out` versions - currently we auto-generate those in a way which expects them to exist.

File: lazy_tensor_core/test/cpp/cpp_test_util.cpp [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+  torch::Tensor output = testfn(input_vars);
+  torch::Tensor xoutput = testfn(xinput_vars);
+  AllClose(output, xoutput, rtol, atol);
+
+  std::vector<torch::Tensor> outs = {output};
+  std::vector<torch::Tensor> xouts = {xoutput};
+  for (int d = 1; d <= derivative_level; ++d) {
+    // Check grad of sum(outs) w.r.t inputs_w_grad.
+    torch::Tensor sum = torch::zeros_like(outs[0]).sum();
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/test/cpp/cpp_test_util.cpp:262]: It is, because we want to propagate gradients. If we just instantiate a fresh zero, it doesn't have any gradient history.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_xla_type.cpp [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+  return bridge::AtenFromLtcTensor(LazyTensor::gelu_backward(
+      bridge::GetLtcTensor(grad), bridge::GetLtcTensor(self)));
+}
+
+at::Tensor AtenXlaType::max_pool2d(const at::Tensor& self,
+                                   at::IntArrayRef kernel_size,
+                                   at::IntArrayRef stride,
+                                   at::IntArrayRef padding,
+                                   at::IntArrayRef dilation, bool ceil_mode) {
+  return aten_autograd_ops_ts::MaxPool2dAutogradFunctionTS::apply(
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_xla_type.cpp:341]: For max pool, we need to save the indices to compute the backward gradient. So, when we get called with `max_pool2d` and expected to compute gradients later, we need to do this special override to set aside the indices.

I'm pretty sure the core used to call `max_pool2d_with_indices` unconditionally a while ago, then it got changed to optimize for use cases in which people didn't want the indices and didn't want gradients either - think `with torch.no_grad()` and calling `max_pool2d` from user code.

File: lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_xla_type.cpp [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+  return std::tuple<at::Tensor, at::Tensor, at::Tensor>(
+      bridge::CreateLtcTensor(std::get<0>(x_result),
+                              bridge::GetLtcDevice(grad_output)),
+      bridge::CreateLtcTensor(std::get<1>(x_result),
+                              bridge::GetLtcDevice(grad_output)),
+      bridge::CreateLtcTensor(std::get<2>(x_result),
+                              bridge::GetLtcDevice(grad_output)));
+}
+
+at::Tensor AtenXlaType::convolution_overrideable(
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/aten_xla_type.cpp:198]: It's quite bad as-is, I'm calling the `slow_conv*` API for simplicity. We'll need to replicate the algorithm selection from core, but that's about it.

File: lazy_tensor_core/test/cpp/cpp_test_util.cpp [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+void MaybeDumpGraph(const at::Tensor& tensor) {
+  static std::string dump_graph =
+      lazy_tensors::sys_util::GetEnvString("XLA_TEST_DUMP_GRAPHS", "");
+  if (!dump_graph.empty() && bridge::IsLtcTensor(tensor)) {
+    std::string graph_str;
+    if (dump_graph == "text") {
+      graph_str = GetTensorTextGraph(tensor);
+    } else if (dump_graph == "hlo") {
```
[the remark this answers is not in the document]
Comment (asuhan) [reply lazy_tensor_core/test/cpp/cpp_test_util.cpp]: Indeed, removed it.
